src/java.base/share/classes/java/net/ServerSocket.java

```diff
@@ -78,18 +78,23 @@
  * @since   1.0
  */
 public class ServerSocket implements java.io.Closeable {
-    /**
-     * The underlying SocketImpl
-     */
+    // the underlying SocketImpl
     private final SocketImpl impl;
 
-    /**
-     * Various states of this socket, need stateLock to change.
-     */
+    // various states
     private volatile boolean created;   // impl.create(boolean) called
     private volatile boolean bound;
     private volatile boolean closed;
-    private final Object stateLock = new Object();
+
+    // used to coordinate creating and closing underlying socket
+    private final Object socketLock = new Object();
+
+    /**
+     * Creates a server socket with the given {@code SocketImpl}.
+     */
+    private ServerSocket(Void unused, SocketImpl impl) {
+        this.impl = Objects.requireNonNull(impl);
+    }
 
```

Comment:
What's the reason for adding this new private constructor? I don't see what the problem with the original single arg one is. `checkPermission` could be declared to return void instead of Void though.

Reply:
Permission checks in constructors of non-final classes are problematic when the subclass has a finalizer. It's fixed here with the same idiom that we use in other places. The return is used by as parameter so it has to be Void.

Reply:
Okay, it just seemed odd that `checkPermission` already had return type Void as if a change like this was started but not completed.

Reply:
Yeah, this idiom might be confusing when you see it first time. The checkPermission method could return another type, we've just used Void to make it a bit clearer that the parameter is not used.